zrml/shares/src/lib.rs

```diff
@@ -1,313 +1,152 @@
-        /// pallet.
-        #[weight = 0]
-        pub fn wrap_native_currency(origin, amount: BalanceOf<T>) {
-            let sender = ensure_signed(origin)?;
-
-            Self::do_wrap_native_currency(sender, amount)?;
-        }
-
-        #[weight = 0]
-        pub fn unwrap_native_currency(origin, amount: BalanceOf<T>) {
-            let sender = ensure_signed(origin)?;
-
-            Self::do_unwrap_native_currency(sender, amount)?;
-        }
-    }
-}
-
-impl<T: Trait> Module<T> {
-    pub fn set_balance(
-        share_id: T::Hash,
-        who: &T::AccountId,
-        balance: BalanceOf<T>,
-    ) -> DispatchResult {
-        <Accounts<T>>::mutate(share_id, who, |data| data.free = balance);
-        Ok(())
-    }
-
-    pub fn set_reserved(
-        share_id: T::Hash,
-        who: &T::AccountId,
-        reserved: BalanceOf<T>,
-    ) -> DispatchResult {
-        <Accounts<T>>::mutate(share_id, who, |data| data.reserved = reserved);
-        Ok(())
-    }
-
-    #[inline]
-    fn get_module_id() -> T::AccountId {
-        T::ModuleId::get().into_account()
     }
 }
 
-impl<T: Trait> Shares<T::AccountId, T::Hash> for Module<T> {
+impl<T> Shares<T::AccountId> for Module<T>
+where
+    T: Trait,
+{
     type Balance = BalanceOf<T>;
+    type Hash = T::Hash;
 
-    fn free_balance(share_id: T::Hash, who: &T::AccountId) -> BalanceOf<T> {
-        Self::accounts(share_id, who).free
+    fn free_balance(asset: Asset<Self::Hash>, who: &T::AccountId) -> BalanceOf<T> {
+        T::Currency::free_balance(asset, who)
     }
 
-    fn total_supply(share_id: T::Hash) -> BalanceOf<T> {
-        <TotalSupply<T>>::get(share_id)
+    fn total_supply(asset: Asset<Self::Hash>) -> BalanceOf<T> {
+        T::Currency::total_issuance(asset)
     }
 
-    fn destroy(share_id: T::Hash, from: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
-        if amount.is_zero() {
-            return Ok(());
-        }
-
-        Self::ensure_can_withdraw(share_id, from, amount)?;
-
-        <TotalSupply<T>>::mutate(share_id, |am| *am -= amount);
-        Self::set_balance(share_id, from, Self::free_balance(share_id, from) - amount)?;
-
+    fn destroy(
+        asset: Asset<Self::Hash>,
+        from: &T::AccountId,
+        amount: BalanceOf<T>,
+    ) -> DispatchResult {
+        T::Currency::slash(asset, from, amount);
         Ok(())
     }
 
-    fn destroy_all(share_id: T::Hash) -> DispatchResult {
-        <Accounts<T>>::remove_prefix(share_id);
-        <TotalSupply<T>>::remove(share_id);
-
+    fn destroy_all(asset: Asset<Self::Hash>) -> DispatchResult {
+        for account_id in T::Currency::accounts_by_asset(&asset) {
+            let reserved_balance = T::Currency::reserved_balance(asset, &account_id);
+            T::Currency::unreserve(asset, &account_id, reserved_balance);
+            let free_balance = T::Currency::free_balance(asset, &account_id);
+            T::Currency::slash(asset, &account_id, free_balance);
```

Comment:
I think what we want to do is instead add another method to the super trait that can access the `remove` method on the storage value directly so we just wipe it clean from storage.

Reply:
Thanks. Just moved `destroy_all` logic into `ZeitgeistMultiReservableCurrency`.